### ISSUE TITLE
SoCo erroring on import

### ISSUE THREAD

**Issue by peterwallhead** [SoCo/SoCo#918]
- Raspberry Pi
- Python 3.9.2

```
Jun 27 05:31:30 raspberrypi python3[743]:     import soco
Jun 27 05:31:30 raspberrypi python3[743]:   File "/usr/local/lib/python3.9/dist-packages/soco/__init__.py", line 12, in <module>
Jun 27 05:31:30 raspberrypi python3[743]:     from .core import SoCo
Jun 27 05:31:30 raspberrypi python3[743]:   File "/usr/local/lib/python3.9/dist-packages/soco/core.py", line 55, in <module>
Jun 27 05:31:30 raspberrypi python3[743]:     from .zonegroupstate import ZoneGroupState
Jun 27 05:31:30 raspberrypi python3[743]:   File "/usr/local/lib/python3.9/dist-packages/soco/zonegroupstate.py", line 66, in <module>
Jun 27 05:31:30 raspberrypi python3[743]:     from lxml import etree as LXML
Jun 27 05:31:30 raspberrypi python3[743]: ImportError: libxslt.so.1: cannot open shared object file: No such file or directory
```

My first thought is perhaps I've missed installing a new dependency? 

Open to all suggestions please, as this has only just started happening as of today (27/06/2022).

Thank you!

**Comment by peterwallhead** [comment]
Ok, a quick google search tells me that it **was** a missing dependency. Still odd that this has only just started happening though.

New/recent release of SoCo?

Running:
```sudo apt-get install libxslt-dev```
seems to have resolved this.

**Comment by WillyChuck** [comment]
This might have sufficed?

sudo apt install python3-lxml

**Comment by peterwallhead** [comment]
Perhaps @WillyChuck, happy to try it out on the next run. Thank you.

**Comment by pwt** [comment]
Hi folks: `lxml` was introduced as a new requirement a week ago in [SoCo v0.28.0](https://github.com/SoCo/SoCo/releases/tag/v0.28.0). It's a binary requirement, and there's some discussion of that in the associated [PR](https://github.com/SoCo/SoCo/pull/913#issuecomment-1120164450).

When SoCo is installed or updated using `pip`, it should automatically install `lxml`, and there are pre-built binary wheels for most of the platforms I've tested, including the Pi. (The only platform I've seen so far on which the wheel has to be built is macOS running on Apple Silicon.) I haven't encountered any situations where I've needed to install system libraries.

@peterwallhead Could you comment on how you're installing SoCo and the full details of your platform, in case we've missed something here?

**Comment by pwt** [comment]
When you install `lxml` just using `pip install lxml`, does it download a pre-built binary or does it perform the build locally?

**Comment by peterwallhead** [comment]
See below for a plain ```pip install lxml```

```
~ $ pip install lxml
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting lxml
  Downloading https://www.piwheels.org/simple/lxml/lxml-4.9.0-cp37-cp37m-linux_armv7l.whl (2.3 MB)
     |████████████████████████████████| 2.3 MB 466 kB/s
Installing collected packages: lxml
Successfully installed lxml-4.9.0
```

**Comment by pwt** [comment]
Thanks for testing. That's pulling down a pre-built binary wheel, as I'd expect.

Looking at the [`lxml` installation instructions](https://lxml.de/installation.html), there's also a dependency on the `libxml2` and `libxslt` system libraries, **if** the pre-built binary is dynamically linked.

The Windows binary is statically linked, so has no additional dependencies.

Based on the binary sizes, I think the ARM 64 binary (6.6MB) is statically linked:
```
Using cached lxml-4.9.0-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl (6.6 MB)
```
whereas the ARM 32 binary (2.3MB) is dynamically linked:
```
Downloading https://www.piwheels.org/simple/lxml/lxml-4.9.0-cp37-cp37m-linux_armv7l.whl (2.3MB)
```
For systems which obtain a dynamically linked `lxml` binary the required system libraries will need to be present or separately installed. There are notes on how to do this in the [`lxml` documentation](https://lxml.de/installation.html#where-to-get-it).

I'll see what other reports of installation issues emerge, and I'll update the release notes to reference this issue.

**Comment by KennethNielsen** [comment]
@pwt Maybe we should trap the import error and give a better error message with installation instructions etc?

**Comment by pwt** [comment]
> @pwt Maybe we should trap the import error and give a better error message with installation instructions etc?

Sounds like a good idea. Somehow I knew this was going to be trouble :)

**Comment by KennethNielsen** [comment]
@pwt well, I think it was difficult to make sure but you did try. I'm actually wondering whether a revert or rewrite around lxml would be in order. It is a really strong selling point that installation is just `pip install` and you are done. We should get the author in on that discussion.

**Comment by jjlawren** [comment]
Honestly, as `lxml` is well-maintained and very commonly used in other XML-dependent libraries, I'd suggest to go further and continue migration over from `ElementTree`. That was [one of the thoughts](https://github.com/SoCo/SoCo/pull/913#issuecomment-1120752703) back in the original PR. The install issues can be improved and docs can be updated for the edge-cases where the install is not seamless.

**Comment by pwt** [comment]
In #920 I've added a descriptive error message in the event of an `lxml` `ImportError`, and a clean exit. Comments in the PR, please.

**Comment by KennethNielsen** [comment]
@jjlawren I think we should move that discussion elsewhere. I will see about opening an issue for the discussion.

**Comment by pwt** [comment]
I've merged #920. This improvement will appear in v0.28.1 (from the [v0.28 fixes](https://github.com/SoCo/SoCo/tree/v0.28.fixes) branch) at some point.

Now closing this issue in favour of #921.
